Compress STDOUT if redirected to file with a compression extension

v01dXYZ opened the pull request:
The code to get the file path from a file descriptor was already in `FormatFactory.cpp`. It is now moved to a specific function in `Common` to make it available to `ClientBase` too.

Tell me if you think the test is enough as I am not well versed at all with your QA procedures. 

Fixes #63496 

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

On Linux and MacOS, if the program has STDOUT redirected to a file with a compression extension, use the corresponding compression method instead of nothing (making it behave similarly to `INTO OUTFILE` ). 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features).
  I don't know where to put this documentation in `docs/`.
  Proposition:
  When using a CLI client on Linux and MacOS, redirecting STDOUT to a file with a compression extension automatically compresses the output with the corresponding compression method.

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

<details>
    <summary>Modify your CI run</summary>

**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ] <!---ci_include_fast--> Fast test
- [ ] <!---ci_include_integration--> Integration Tests
- [ ] <!---ci_include_stateless--> Stateless tests
- [ ] <!---ci_include_stateful--> Stateful tests
- [ ] <!---ci_include_unit--> Unit tests
- [ ] <!---ci_include_performance--> Performance tests
- [ ] <!---ci_include_asan--> All with ASAN
- [ ] <!---ci_include_tsan--> All with TSAN
- [ ] <!---ci_include_analyzer--> All with Analyzer
- [ ] <!---ci_include_azure --> All with Azure
- [ ] <!---ci_include_KEYWORD--> Add your option here

#### Exclude tests:
- [ ] <!---ci_exclude_fast--> Fast test
- [ ] <!---ci_exclude_integration--> Integration Tests
- [ ] <!---ci_exclude_stateless--> Stateless tests
- [ ] <!---ci_exclude_stateful--> Stateful tests
- [ ] <!---ci_exclude_performance--> Performance tests
- [ ] <!---ci_exclude_asan--> All with ASAN
- [ ] <!---ci_exclude_tsan--> All with TSAN
- [ ] <!---ci_exclude_msan--> All with MSAN
- [ ] <!---ci_exclude_ubsan--> All with UBSAN
- [ ] <!---ci_exclude_coverage--> All with Coverage
- [ ] <!---ci_exclude_aarch64--> All with Aarch64
- [ ] <!---ci_exclude_KEYWORD--> Add your option here

#### Extra options:
- [ ] <!---do_not_test--> do not test (only style check)
- [ ] <!---no_merge_commit--> disable merge-commit (no merge from master before tests)
- [ ] <!---no_ci_cache--> disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ] <!---batch_0--> 1
- [ ] <!---batch_1--> 2
- [ ] <!---batch_2--> 3
- [ ] <!---batch_3--> 4

<details>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=63662) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=63662) before we can accept your contribution.<br/><hr/>**v01dxyz** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=63662) it.</sub>

robot-clickhouse commented:
<!-- automatic status comment for PR #63662 from v01dXYZ/ClickHouse:63496-compression-from-file-descriptor -->
*This is an automated comment for commit 377dcff68958633f23ff965b951a13a28153c97c with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63662/377dcff68958633f23ff965b951a13a28153c97c/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

v01dXYZ committed:
Compress STDOUT if redirected to file with a compression extension

* Add a new member to ClientBase: default_output_compression_method
* Move the code to get file path from file descriptor to a separate
  Common function.

The stateless test is almost a copy-paste of 02001_compress_output_file.

Fixes https://github.com/ClickHouse/ClickHouse/issues/63496

Avogar commented:
Failed tests are just flaky and not related to the changes in this PR

Avogar reviewed (approved):
Great! Thanks a lot for the contribution!

clickgapai commented:
Hi @v01dXYZ @Avogar — while reviewing this PR I found the following:

- **Bug (P2):** STDOUT compression not applied when output format is explicitly specified — https://github.com/ClickHouse/ClickHouse/issues/104441

Happy to discuss — close anything that's wrong or already addressed.
